source/common/router/scoped_rds.cc

```diff
@@ -125,15 +125,94 @@ ScopedRdsConfigSubscription::RdsRouteConfigProviderHelper::RdsRouteConfigProvide
     ScopedRdsConfigSubscription& parent, std::string scope_name,
     envoy::extensions::filters::network::http_connection_manager::v3::Rds& rds,
     Init::Manager& init_manager)
-    : parent_(parent), scope_name_(scope_name),
-      route_provider_(std::dynamic_pointer_cast<RdsRouteConfigProviderImpl>(
-          parent_.route_config_provider_manager_.createRdsRouteConfigProvider(
-              rds, parent_.factory_context_, parent_.stat_prefix_, init_manager))),
+    : parent_(parent), scope_name_(scope_name), on_demand_(false) {
+  initRdsConfigProvider(rds, init_manager);
+}
+
+ScopedRdsConfigSubscription::RdsRouteConfigProviderHelper::RdsRouteConfigProviderHelper(
+    ScopedRdsConfigSubscription& parent, std::string scope_name)
+    : parent_(parent), scope_name_(scope_name), on_demand_(true) {
+  parent_.stats_.on_demand_scopes_.inc();
+}
+
+void ScopedRdsConfigSubscription::RdsRouteConfigProviderHelper::addOnDemandUpdateCallback(
```

Comment:
I spent a few mins staring at this functions and still find it a bit confusing. Can you decompose it, or simplify it? Even better comments to make clear why the different steps during initialization and orchestration have to happen and how they related to the big picture, assuming the reader isn't an expert on RDS/SRDS init?

Reply:
I will explain why we need different constructor when eager loading scopes and on demand scopes are used in the comment.  Also, this function can be decomposed if we want to.